### DISCUSSION TITLE
Calculation step 2 comments

### DISCUSSION THREAD

**Discussion by TcM1911** [NextronSystems/gimphash#3] [Ideas]
Step 4:
> If the part before the first / contains a . and is NOT in the following list, ignore the function name: (ignoring private repositories; often serve as source code instead of 'imports' that we'd like to hash here)

Restricting to this specific list may result in a lot of legitimate 3rd party packages getting excluded from the calculation. For example [bbolt](https://github.com/etcd-io/bbolt#importing-bbolt), a popular database would. While it is hosted on Github, the documentation says to use `import bolt "go.etcd.io/bbolt"`.

The alternative option of using file path may be hard to "codify" in a specification. Would we not need to define an algorithm for deciding if the package should be included or not as part of the specification to ensure different implementations returns the same hash for the same file? I've tried to implement something like this for [gore](https://github.com/goretk/gore) but it's not 100% accurate. This is because there isn't a standard folder structure for Go source code. Handling formats used right now is possible but what do we do for future source code structures? You also have issues with obfuscators that removes the file path information.

An alternative is to use buildinfo from the binary to identify which packages are part of the main module. While it should be very accurate, this information is not always available... It can exist in binaries compiled with 1.13 or newer but it is possible to exclude it.

With these possible issues, would it be better to just hash all packages to ensure consistency?

**Comment by secDre4mer** [comment]
> The alternative option of using file path may be hard to "codify" in a specification. Would we not need to define an algorithm for deciding if the package should be included or not as part of the specification to ensure different implementations returns the same hash for the same file? I've tried to implement something like this for [gore](https://github.com/goretk/gore) but it's not 100% accurate. This is because there isn't a standard folder structure for Go source code. Handling formats used right now is possible but what do we do for future source code structures? You also have issues with obfuscators that removes the file path information.

This was pretty much the reasoning why the exclusion is as it is currently. Your work on gore was verify helpful in developing the gimphash so far, so thank you for that!

Obfuscators like [garble](https://github.com/burrowers/garble/) also affect vendored and stdlib package names, so I'd ignore them for now. Gimphash results from them are not helpful in any case.

As you've pointed out, filepath and buildinfo information are possibly not fully reliable or not resistant enough to change to be used as part of an algorithm, and I've therefore tried to avoid using them so far. We'd also need to parse more internal data structures, which makes us more vulnerable to future changes to these structures (after all, the pclntab structure was already modified twice in recent history).

Regarding hashing all packages vs. hashing only the current whitelist (or extending that whitelist): I agree that hashing all packages would ensure better consistency / make the hash more specific. The main drawback from my perspective is that it also causes private packages to be included: If a software adds a new package that it uses internally, its gimphash might change, even though its dependencies do not. 
I'm not sure yet what is the better approach.
